src/components/ColorRampDropdown.tsx

```diff
@@ -58,46 +84,93 @@ const ColorRampSelector: React.FC<ColorRampSelectorProps> = (propsInput): ReactE
-      // Show the name of the color ramp in the tooltip, but use its internal key for callbacks.
-      const [key, colorRampData] = colorRampEntries[i];
+    for (let i = 0; i < colorRampData.length; i++) {
       contents.push(
-        <Tooltip title={colorRampData.name} placement="right" key={key} trigger={["hover", "focus"]}>
-          <Button key={key} onClick={() => props.onChange(key, false)} rootClassName={styles.dropdownButton}>
-            <img src={colorRampData.colorRamp.createGradientCanvas(120, theme.controls.height).toDataURL()} />
+        <Tooltip title={colorRampData[i].name} placement="right" key={i} trigger={["hover", "focus"]}>
+          <Button onClick={() => onClick(colorRampData[i])} rootClassName={styles.dropdownButton}>
+            <img src={colorRampData[i].colorRamp.createGradientCanvas(120, theme.controls.height).toDataURL()} />
           </Button>
         </Tooltip>
       );
     }
     return contents;
-  }, [props.colorRamps]);
+  };
+
+  /** The contents of the dropdown that appears when you hover over the button */
+  const dropdownContents = useMemo(() => {
+    if (props.useCategoricalPalettes) {
+      // Make categorical palettes by converting them into color ramps.
+      const onClick = (paletteData: ColorRampData): void => {
+        const colors = props.categoricalPalettes.get(paletteData.key)?.colors;
+        if (colors) {
+          props.onChangePalette(colors);
+        }
+      };
+      // Generate color ramps from the palettes.
+      const paletteData = Array.from(props.categoricalPalettes.values());
+      // Append the missing ColorRamp into the palette data so it can be handled as a ColorRampData.
+      const colorRampData = paletteData.map((data) => {
+        const visibleColors = data.colors.slice(0, Math.max(1, props.numCategories));
+        return { ...data, colorRamp: new ColorRamp(visibleColors, ColorRampType.HARD_STOP) };
+      });
+      return makeRampButtonList(colorRampData, onClick);
+    } else {
+      // Make gradient ramps instead
+      return makeRampButtonList(Array.from(props.colorRamps.values()), (rampData) => {
+        props.onChangeRamp(rampData.key, false);
+      });
+    }
+  }, [props.colorRamps, props.useCategoricalPalettes, props.categoricalPalettes, props.numCategories]);
 
   /// Rendering
 
+  // Swap between two image sources as needed
+  const dropdownButtonImgSrc = props.useCategoricalPalettes ? paletteImgSrc : rampImgSrc;
+
   const buttonDivClassName = styles.buttonContainer + (props.disabled ? ` ${styles.disabled}` : "");
-  const dropdownContainerClassName = styles.dropdownContainer + (forceOpen ? ` ${styles.forceOpen}` : "");
+
+  const dropdownClassNames = [styles.dropdownContainer];
+  props.useCategoricalPalettes && dropdownClassNames.push(styles.categorical);
+  forceOpen && dropdownClassNames.push(styles.forceOpen);
+  const dropdownContainerClassName = dropdownClassNames.join(" ");
```

Comment:
Idk how I feel about using lazy `&&` purely for a side effect like this, rather than to produce a value. I get why it looks a lot nicer in this context... but I'd encourage `if` here.